src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -430,6 +415,56 @@ bool CoreWorkerDirectActorTaskSubmitter::IsActorAlive(const ActorID &actor_id) c
   return (iter != client_queues_.end() && iter->second.rpc_client);
 }
 
+void CoreWorkerDirectActorTaskSubmitter::AppendTask(
+    const TaskSpecification &task_spec,
+    const std::function<void(const ActorID &)> &callback) {
+  auto actor_id = task_spec.ActorId();
+  RAY_LOG(DEBUG) << "AppendTask task: " << task_spec.DebugString();
+  RAY_CHECK(task_spec.IsActorTask());
+  uint64_t send_pos = 0;
+  absl::MutexLock lock(&mu_);
+  auto it = client_queues_.find(actor_id);
+  RAY_CHECK(it != client_queues_.end());
+
+  // We must fix the send order prior to resolving dependencies, which may
+  // complete out of order. This ensures that we will not deadlock due to
+  // backpressure. The receiving actor will execute the tasks according to
+  // this sequence number.
+  send_pos = task_spec.ActorCounter();
+  auto inserted = it->second.requests.emplace(send_pos, std::make_pair(task_spec, false));
+  RAY_CHECK(inserted.second);
+  actor_task_io_service_.post(
+      [this, task_spec, callback, actor_id]() {
+        RAY_UNUSED(SubmitTask(task_spec));
+        if (callback) {
+          callback(actor_id);
+        }
+      },
+      "CoreWorker.SubmitActorTask");
+  it->second.cur_pending_calls++;
+}
+
+void CoreWorkerDirectActorTaskSubmitter::WaitForShutdown() {
+  if (actor_task_io_thread_.joinable()) {
+    actor_task_io_thread_.join();
+  }
+}
+
+void CoreWorkerDirectActorTaskSubmitter::Shutdown() { actor_task_io_service_.stop(); }
+
+void CoreWorkerDirectActorTaskSubmitter::RunActorTaskIOService() {
+  SetThreadName("actor_task");
+  actor_task_io_service_.run();
+}
+
+bool CoreWorkerDirectActorTaskSubmitter::FullOfPendingTasks(const ActorID &actor_id) {
+  absl::MutexLock lock(&mu_);
+  auto it = client_queues_.find(actor_id);
+  RAY_CHECK(it != client_queues_.end());
```

Comment:
is this too aggressive?

Reply:
I am not sure, it seems to be a general practice in ray core code.